keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md

```diff
@@ -118,78 +119,79 @@ Currently, kubeadm vendors in pieces of Go code from the components to be able t
+The loaded configs, that were supplied with `--component-config-upgrades` would be taken into account when printing the additional component config information.
 
-Kustomize patches will be applied on both user supplied and kubeadm generated core component configs, just before they are stored in the local files from which the config will be consumed by the components. In the case of user supplied core component configs, neither the Kustomize patches, nor the result after applying them is going to be stored on the cluster. The config maps are going to contain only the vanilla user supplied component configs.
+### "kubeadm upgrade apply" changes
 
-If a user has provided Kustomize patches for a config version, that is different from the current one, kubeadm is going to bail out with an error, prompting the user to either delete the old patches, or to upgrade them to the new version. This is best implemented using a kubeadm pre-flight check.
+As supported component configs are expected to increase in number in the future and some of them might require manual user migration prior to upgrade, a handy and secure method for supplying new configs is necessary.
 
-UX wise, this feature is going to be reusing the existing kubeadm options that were introduced with the implementation of the kubeadm advanced configurations KEP. This allows for kustomizations to be applied to component configs on a per-node basis during init, join and upgrade just before the config is supposed to be stored on the local node file.
+For that matter `kubeadm upgrade apply` also gains the same `--component-config-upgrades` flag and with the same semantics as in `kubeadm upgrade plan`.  
+In addition to that, the newly used configs, from the file supplied with `--component-config-upgrades`, would be stored in the cluster as user specified configs during the upgrade process.
 
-### Stop using component config internal types
+### Introduce the "kubeadm alpha component-configs print-upgradeable" command
 
-As kubeadm will no longer default, validate or migrate component configs inside its code base, the need to use component config internal types is no longer needed. In that case the usage of internal types can be removed. This would also allow us to reduce the number of imported packages and to drop one of the last dependencies on the kubernetes/kubernetes code base.  
-The switch is going to be performed by replacing each internal type with a corresponding versioned type. This, by itself, is not going to have any impact on users.
+This new command would fetch the YAML documents of any component configs that are in use in the cluster and require manual migration to a newer version prior to cluster upgrade. Then it would print those YAML documents to the standard output or to a file, specified with the `--output` command line option.  
+If there are no configs that need manual upgrade, this command would print nothing.
```

Comment:
wondering how would it work under the hood? if the user is using the latest kubeadm 1.Y binary vs kubeadm binary 1.Y-1.
i'm assuming they need to use 1.Y to tell them what has to be upgraded, but how would the 1.Y binary know if the config in the cluster version 1.Y-1 needs an upgrade?

does that mean that the 1.Y binary needs to know about the component config defaults for 1.Y and 1.Y-1, so that it can compare those to the SHA in the 1.Y-1 cluster?

Reply:
The detection mechanism is not much different than the one used for `kubeadm upgrade plan` (see kubernetes/kubernetes#88124).
In essence, kubeadm would check the groups. If the API group matches, but the version is not the supported one - then the config needs upgrading.
In that case, it's just printed as is (for this particular command).

Hence, you should always use the 1.Y binary.

As for the SHAs - we'll use them to check if a generated config was modified or not.

Reply:
but the 1.Y binary still needs to know about generated config 1.Y and 1.Y-1, i guess? not that this is much of an issue, as far as i can tell.